File: go/programmers/타겟_넘버.go
package targetNumber

func solution(numbers []int, target int) int {
	visited := make([]bool, len(numbers))

	result := dfs(numbers, target, 0, 0, visited)

	return result
}

func dfs(numbers []int, target int, total int, cur int, visited []bool) int {
	if cur == len(numbers) {
		if total == target {
			return 1
		}
		return 0
	}

	visited[cur] = true

	number := numbers[cur]

	minusTotal := total - number
	plusTotal := total + number

	return dfs(numbers, target, minusTotal, cur+1, visited) + dfs(numbers, target, plusTotal, cur+1, visited)
}

File: java/programmers/타겟_넘버.java
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) {
        int answer = 0;

        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] {0, 0});

        while(!queue.isEmpty()) {
            int[] state = queue.poll();
            int index = state[0];
            int total = state[1];

            if (index >= numbers.length) {
                if (total == target) {
                    answer++;
                }
            } else {
                queue.add(new int[] {index + 1, total + numbers[index]});
                queue.add(new int[] {index + 1, total - numbers[index]});
            }
        }

        return answer;
    }

//    public int solution(int[] numbers, int target) {
//        return dfs(numbers, target, 0, 0);
//    }

    private int dfs(int[] numbers, int target, int index, int total) {
        if (index == numbers.length) {
            return total == target ? 1 : 0;
        }

        return dfs(numbers, target, index + 1, total + numbers[index])
                + dfs(numbers, target, index + 1, total - numbers[index]);
    }
}